### ISSUE TITLE
Build fails with `"PreferredUILang" parameter is not supported by the "VsTsc"`

### ISSUE THREAD

**Issue by mhegazy** [Microsoft/TypeScript#14801]
After installing TS 2.2 for VS 2015, build fails in VS 2017 with error:

```
MSB4064	The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property.
```

More details can be found at https://vsfeedback/comment/826493

**Comment by jkonecki** [comment]
How was this issue fixed? I experience it with TS 2.2.2 building on VSTS (Hosted)

**Comment by mhegazy** [comment]
@zhengbli can you look into the vsts issue?

**Comment by jkonecki** [comment]
I get `packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(204,7): Error MSB4064: The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property.` while building the solution.

The targets file is still setting `PreferredUILang`:

```
  <Target Name="PreComputeCompileTypeScriptWithTSConfig"
          Condition="'@(ConfigFiles)' != ''">
    <VsTsc
      ToolPath="$(TscToolPath)"
      ToolExe="$(TscToolExe)"
      PreferredUILang="$(PreferredUILang)"
      TSConfigFile="%(ConfigFiles.Identity)"
      YieldDuringToolExecution="$(TscYieldDuringToolExecution)"
      ProjectDir="$(TypeScriptProjectDir)"
      ToolsVersion="$(TypeScriptToolsVersion)"
      TypeScriptCompileBlocked="$(TypeScriptCompileBlocked)"
      ComputeOutputOnly="true">

      <Output TaskParameter="GeneratedJavascript" ItemName="emittedFiles" />
    </VsTsc>
    <ItemGroup>
      <GeneratedJavascript Include="@(emittedFiles)" KeepDuplicates="false" />
    </ItemGroup>
  </Target>
```

**Comment by jkonecki** [comment]
I can confirm that removing all instances of `PreferredUILang="$(PreferredUILang)"` from the `Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets` solves the problem.
Since there are no code changes related to this issue I can only assume it was closed without actuak fix.

**Workaround**: Remove all instances of `PreferredUILang="$(PreferredUILang)"` from the `packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets` file and include the package folder into source control.

**Comment by mhegazy** [comment]
Can you share the version of the tasks dll?

**Comment by jkonecki** [comment]
The version of _TypeScript.Tasks.dll_ included in _Microsoft.TypeScript.MSBuild.2.2.2.nupkg_ is *14.40321.1.0*

**Comment by zhengbli** [comment]
@jkonecki can you try set the build output verbosity level to diagnostics by going to tools -> options -> Project and Solutions -> Build and run, and set 
![image](https://cloud.githubusercontent.com/assets/1171301/24368034/cc1958de-12d3-11e7-96e9-c8eb817b0b3e.png)

And then look for the path of "tsc.exe"? I would like to see if the correct one was used.

**Comment by jkonecki** [comment]
I cannot provide the path to `tsc.exe` as the problem is with calling `VsTsc` task in the first place.
Here is the detailed output from the build:

```
2017-03-28T08:20:21.7060412Z ##[debug]Processed: ##vso[task.logdetail id=e5a2f899-a750-4145-b31c-19d9ed8faf18;parentid=;name=packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets;type=Build;starttime=2017-03-28T08:20:21.6940406Z;state=InProgress;]
2017-03-28T08:20:21.7130417Z ##[error]packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(203,7): Error MSB4064: The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property.
2017-03-28T08:20:21.7130417Z ##[debug]Processed: ##vso[task.logissue type=Error;sourcepath=C:\a\1\s\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets;linenumber=203;columnnumber=7;code=MSB4064;]The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property.
2017-03-28T08:20:21.7130417Z ##[debug]Processed: ##vso[task.logdetail id=e5a2f899-a750-4145-b31c-19d9ed8faf18;parentid=;type=Build;result=Failed;finishtime=2017-03-28T08:20:21.6990408Z;progress=100;state=Completed;parentid=;name=;]
2017-03-28T08:20:21.7130417Z C:\a\1\s\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(203,7): error MSB4064: The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property. [C:\a\1\s\Portfolio\Portfolio.csproj]
2017-03-28T08:20:21.7130417Z ##[error]packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(199,5): Error MSB4063: The "VsTsc" task could not be initialized with its input parameters. 
2017-03-28T08:20:21.7140418Z ##[debug]Processed: ##vso[task.logissue type=Error;sourcepath=C:\a\1\s\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets;linenumber=199;columnnumber=5;code=MSB4063;]The "VsTsc" task could not be initialized with its input parameters. 
2017-03-28T08:20:21.7140418Z C:\a\1\s\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(199,5): error MSB4063: The "VsTsc" task could not be initialized with its input parameters.  [C:\a\1\s\Portfolio\Portfolio.csproj]
```

You need to either remove setting of `PreferredUILang` input parameter in the targets file (assuming task is correct) or add support for `PreferredUILang` input parameter to the task (assuming targets file is correct). I suppose the former change should be made.

**Comment by zhengbli** [comment]
Actually can you send the entire log? My email is zhengbli@microsoft.com. Because we need to see if the tasks and the targets have a mismatch, which is likely what has caused the problem. The solution is to align the tasks that support `PreferredUILang` with the targets not containing it, though there are many floating parts and looking at the log could help much.

**Comment by jkonecki** [comment]
I've emailed detailed logs as requested.

> Because we need to see if the tasks and the targets have a mismatch, which is likely what has caused the problem. 

Both task and targets files come from your own NuGet package. You can start by checking them.

**Comment by pablonautic** [comment]
Hello guys,
just FYI,
I landed here because I hit the same error message after installing Microsoft.TypeScript.Compiler 2.2.1 and Microsoft.TypeScript.MSBuild 2.2.2 in an existing project using TS. 

I managed to eliminate it using by removing the leftovers from the previously used TS version, precisely, by manually removing 
`  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.targets" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.targets')" />` somewhere near end of the .csproj file.

So, the first entry 
`  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.Default.props" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.Default.props')" />` was replaced by Nuget during package installation to match the new path, but the second was left as-is and, I think, was causing the issue.

**Comment by jkonecki** [comment]
Thank you @pablonautic for your suggestion.
My problem seems to be a bit different as my project doesn't have `Microsoft.TypeScript.targets` import at all.

**Comment by spustlik** [comment]
I have same problem here in VS2015 14.0.25431.01 Update 3, (VS2017 also installed)
File->New Project->select Typescript/HTML Application with Typescript template.
Compile->OK
install nuget 
```xml
  <package id="Microsoft.TypeScript.MSBuild" version="2.2.2" targetFramework="net452" 
developmentDependency="true" />
```
Compile->FAIL
> ------ Build started: Project: TypeScriptHTMLApp3, Configuration: Debug Any CPU ------
> c:\users\xxx\documents\visual studio 2015\Projects\TypeScriptHTMLApp3\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(204,7): error MSB4064: The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property.
> c:\users\xxx\documents\visual studio 2015\Projects\TypeScriptHTMLApp3\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(200,5): error MSB4063: The "VsTsc" task could not be initialized with its input parameters. 
> ========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

When I deleted all PreferredUILang="$(PreferredUILang)" from .target, it seemed to work.
But my scenario for adding this nuget package was to make it possible to use exact version of TSC, not to tweak it. I need to make it run on hosted team services build also.
My guess is that for VS2015 there is no PreferredUILang prop, but only for VS2017, and there should be two .targets files in nuget package sources.

**Comment by spustlik** [comment]
Downgrading nuget package to version 2.2.1 seems to work. I tried it also in hosted team services build (with option Visual Studio 2017) and it worked

**Comment by zhengbli** [comment]
@spustlik can you try installing the latest TS 2.2 for VS 2015 at http://www.typescriptlang.org/#download-links to see if that solves it
The HTML Application with TypeScript template was using the v1.8 tsc.exe thus caused the issue. Investigating why that happens but you can try that as a workaround

**Comment by spustlik** [comment]
I've already installed this (see image below), this doesn't solve the issue.
My story is that some actual typings from npm was not compatible with older TS (i thing I was on 2.0 before). So I updated VS2015 to use TS 2.2. It worked ok, but my hosted builds on visualstudio.com started to fail, because hosted build environment is using only 1.0-2.0 versions of TS (see log below).
So I started to look for solution to take version of Typescript compiler side-by-side with my VS solution. Nuget package seems to be good solution. As I mentioned above, version 2.2.1 of nuget package is working OK.

2017-03-31T09:33:40.6192505Z ##[section]Starting: Run dir
2017-03-31T09:33:40.6292504Z ==============================================================================
2017-03-31T09:33:40.6292504Z Task         : Command Line
2017-03-31T09:33:40.6292504Z Description  : Run a command line with arguments
2017-03-31T09:33:40.6292504Z Version      : 1.1.2
2017-03-31T09:33:40.6292504Z Author       : Microsoft Corporation
2017-03-31T09:33:40.6292504Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613735)
2017-03-31T09:33:40.6292504Z ==============================================================================
2017-03-31T09:33:40.6571067Z ##[command]dir "C:\Program Files (x86)\Microsoft SDKs\TypeScript\"
2017-03-31T09:33:40.6691049Z  Volume in drive C has no label.
2017-03-31T09:33:40.6691049Z  Volume Serial Number is 78C2-210A
2017-03-31T09:33:40.6691049Z 
2017-03-31T09:33:40.6691049Z  Directory of C:\Program Files (x86)\Microsoft SDKs\TypeScript
2017-03-31T09:33:40.6691049Z 
2017-03-31T09:33:40.6691049Z 12/14/2016  08:02 PM    <DIR>          .
2017-03-31T09:33:40.6691049Z 12/14/2016  08:02 PM    <DIR>          ..
2017-03-31T09:33:40.6691049Z 12/14/2016  08:02 PM    <DIR>          1.0
2017-03-31T09:33:40.6691049Z 12/01/2016  11:59 PM    <DIR>          1.4
2017-03-31T09:33:40.6691049Z 12/02/2016  02:12 AM    <DIR>          1.6
2017-03-31T09:33:40.6691049Z 12/02/2016  02:14 AM    <DIR>          1.7
2017-03-31T09:33:40.6691049Z 12/02/2016  12:47 AM    <DIR>          1.8
2017-03-31T09:33:40.6691049Z 12/02/2016  02:16 AM    <DIR>          2.0
2017-03-31T09:33:40.6691049Z                0 File(s)              0 bytes
2017-03-31T09:33:40.6691049Z                8 Dir(s)  13,663,072,256 bytes free
2017-03-31T09:33:40.6691049Z ##[section]Finishing: Run dir


![image](https://cloud.githubusercontent.com/assets/1135561/24566428/c408c920-1659-11e7-8377-7b35171ca076.png)

**Comment by spustlik** [comment]
If I am trying to find all TypeScript.Tasks.dll, I found it at:
packages\Microsoft.TypeScript.MSBuild.2.2.2
packages\Microsoft.TypeScript.MSBuild.2.2.1
"C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v12.0\TypeScript"\ - 14.31015.1.0

"C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\TypeScript"\ - 14.40221.2.0 
"C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v15.0\TypeScript"\ - 14.40221.2.0

So it seems that older version of TypeScript.Tasks.dll from v14.0 folder is executed instead of that 14.40321.1.0 from 2.2.2 nuget package...

**Comment by zhengbli** [comment]
We actually released a new version (TypeScript 2.2.2) for VS 2015, namely the link. If you can give that a try.

I agree the nuget package should use what comes with the nuget, though it is tied to how the project template was authored (the order of imports etc.) Just to confirm, what are the project templates that have this issue for you? Any others beside the HTML Application one?

**Comment by spustlik** [comment]
Great, this works locally. I overlooked difference beween 2.2.1 and 2.2.2, my apologies.
But when trying to execute hosted build, it fails. 
> 2017-03-31T20:19:19.2361325Z (PreComputeCompileTypeScriptWithTSConfig target) -> 
2017-03-31T20:19:19.2361325Z   C:\a\1\s\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(183,7): error MSB4064: The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property. [C:\a\1\s\SkillDriver.Web\SkillDriver.Web.csproj]
2017-03-31T20:19:19.2361325Z   C:\a\1\s\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets(180,5): error MSB4063: The "VsTsc" task could not be initialized with its input parameters.  [C:\a\1\s\SkillDriver.Web\SkillDriver.Web.csproj]
2017-03-31T20:19:19.2371376Z

**Comment by jkonecki** [comment]
@spustlik so now you're in the same place as me. Try my workaround posted above that involves editing targets file and committing NuGet package folder.

**Comment by mihailik** [comment]
We're having this same issue.

Building from MSBuild command line (coming with VS2015 Update 3), plus TS2.2.2 nuget package (the newer TypeScript VSIX addin is NOT installed).

I understand VSIX is adding value to IDE experience, but nuget should be enough for MSBuild?

**Comment by mihailik** [comment]
Removing these from my CSPROJ solved the problem for us:

```xml
  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.Default.props" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.Default.props')" />

...

  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.targets" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.targets')" />
  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\WebApplications\Microsoft.WebApplication.targets" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\WebApplications\Microsoft.WebApplication.targets')" />
```

I am reluctant to edit anything in `packages` directory, as these edits are hard to reflect across the team consistently.

**Comment by zhengbli** [comment]
The reason is that we have a breaking change between 2.1+ and 1.8 TypeScirpt tasks file, and the nuget package was not enforcing using the matching tasks file. A PR was sent on the managed side to solve this issue.

**Comment by jkonecki** [comment]
Thank you for the update.

Could you please provide the link to the PR?
Will a new NuGet package be provided once the issue is fixed?

**Comment by zhengbli** [comment]
@jkonecki sorry the nuget & visual studio part is not open-sourced yet, which was why you saw this issue gets closed with no changes made to this repo. I'll keep you updated. Yes a new version of the package should be published once this is fixed.

**Comment by mhegazy** [comment]
When adding a nuget package, you will need to **manually** remove the imports to the machine-wide targets and props. having both leads to unknown load order or targets/tasks, which would manifest in such errors.

In Nuget v2 we used to have support for `install.ps1` scripts that removes the machine-wide target/prop imports when the package is installed, but that has been deprecated in [Nuget v3](https://docs.microsoft.com/en-us/nuget/release-notes/nuget-3.1#deprecated) (see more at https://github.com/NuGet/Home/issues/1020).

also related issue is https://github.com/Microsoft/TypeScript/issues/12527

**Comment by shyamal890** [comment]
@mhegazy I manually commented out

```
 <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.Default.props" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.Default.props')" />

  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.targets" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.targets')" />
  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\WebApplications\Microsoft.WebApplication.targets" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\WebApplications\Microsoft.WebApplication.targets')" />

```

Though this doesn't solve the issue. The Hosted Visual Studio Team Services gives the following error:

> Error MSB4064: The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property.

Now, this points me to what @jkonecki suggested earlier, manually removing references to `PreferredUILang`. So apart for commenting out the above code from `.csproj` we also need to remove references to `PreferredUILang` ?

**Comment by jkonecki** [comment]
I think so. Give it a go.

**Comment by shyamal890** [comment]
Frankly, this is too much of a hack. Not only do the user have to manually remove `PreferredUILang` references, he would also have to include the package folder to source control. Again if the VSTS hosted build is set to reinstall nuget packages on build, there is no guarantee that the modified file wouldn't be over-written @mhegazy hope a solution to this is found soon.

Even better if VSTS hosted build supports 2.2.2 as it would remove the need for any kind of changes.

**Comment by jkonecki** [comment]
You don't need to include package folder into the source control. Just
removing PreferedUILang and invalid imports should be enough.

**Comment by shyamal890** [comment]
But the project is built with VSTS hosted build so I would have to include the modified target file to source control

**Comment by mhegazy** [comment]
What version of TS are you using to build?

**Comment by shyamal890** [comment]
@mhegazy 2.2.2.

**Comment by chrismbarr** [comment]
I'm having the exact same problem.  I am attempting to upgrade a TS 1.8.x app to 2.2.x _(since 2.3.x currently has [this issue that I'm attempting to avoid](https://github.com/Microsoft/TypeScript/issues/15536) by waiting to upgrade)_.

So far I've installed the   `Microsoft.TypeScript.Compiler` package at `2.2.1` and the 
 `Microsoft.TypeScript.MSBuild` package at `2.2.2`, and added a `tsconfig.json` file.  It builds locally fine, but when I trigger a build on our TFS server I get the following errors

```
Source\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets (183, 7)
The "PreferredUILang" parameter is not supported by the "VsTsc" task. Verify the parameter exists on the task, and it is a settable public instance property.
Source\packages\Microsoft.TypeScript.MSBuild.2.2.2\tools\microsoft.TypeScript.targets (180, 5)
The "VsTsc" task could not be initialized with its input parameters.
```

I tried following the advice here and I removed the following imports from my `csproj` file
```
<Import Project="$(VSToolsPath)\WebApplications\Microsoft.WebApplication.targets" Condition="'$(VSToolsPath)' != ''" />
<Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.targets" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.targets')" />
<Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.Default.props" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\TypeScript\Microsoft.TypeScript.Default.props')" />
```
I still get the same error.  We have other applications that are on TS 2,2.x and I can't figure out what is different about those apps vs, this one.  Those work without error, but this one does not.

**Comment by chrismbarr** [comment]
ok, this is what worked for me, I took it from: http://stackoverflow.com/a/43933100/79677

1. Uninstall from the project the nuget packages "Microsoft.TypeScript.Compiler" and Microsoft.Typescript.MSBuild"
2. Close VS (to guarantee the real clean of such package)
3. Edit you .csproj file commenting out the imports to the machine-wide targets and props: (all <Import Project="$(MSBuildExtensionsPath32)\Microsoft\...
4. Restart VS, open the project and add again the nuget packages for Typescript. This point correctly updates your .csproj with the correct imports and configuration of the TypeScript compiler.
5. Just for a clean safe: restart VS.

After that, I also made sure that my project file used the correct matching Visual Studio version everywhere.  Here's what a diff looked like when I made this change.

![image](https://cloud.githubusercontent.com/assets/463685/26256937/a8aca1de-3c8c-11e7-9daf-5ef525e8c2aa.png)

We are using VS2015 everywhere and have TFS set to build with that version, so I set the version to be `14.0` everywhere.  This seemed to be the last key!
